## Tracks data analysis

Let's assert tracks data analysis by taking a first inspection of the track dataset

In [1]:
import pandas as pd
import altair as alt
from os import path
import numpy as np
import re
import spacy
import warnings
from langdetect import detect, DetectorFactory
from langdetect.lang_detect_exception import LangDetectException

dataset_path = path.join('..', 'dataset', 'tracks.csv')
df = pd.read_csv(dataset_path, sep=',')

DetectorFactory.seed = 0

def load_spacy_models():
    models_tokenizers = {}
    models_pos_taggers = {}
    
    #lightweight spacy models
    try:
        nlp_it_light = spacy.load('it_core_news_sm', 
                                  disable=['parser', 'tagger', 'ner', 'lemmatizer'])
        nlp_it_light.add_pipe('sentencizer')
        models_tokenizers['it'] = nlp_it_light
        
        nlp_en_light = spacy.load('en_core_web_sm', 
                                  disable=['parser', 'tagger', 'ner', 'lemmatizer'])
        nlp_en_light.add_pipe('sentencizer')
        models_tokenizers['en'] = nlp_en_light
        
    except (IOError, ImportError):
        print("Error: spaCy lightweight models not found")
        return None, None

    #POS tagging models (slower since we keep tagger and parser enabled)
    try:
        models_pos_taggers['it'] = spacy.load('it_core_news_sm', 
                                             disable=['ner', 'lemmatizer'])
        
        models_pos_taggers['en'] = spacy.load('en_core_web_sm', 
                                             disable=['ner', 'lemmatizer'])
        
    except (IOError, ImportError):
        print("Error: spaCy POS models not found")
        return None, None

    return models_tokenizers, models_pos_taggers

models_tokenizers, models_pos_taggers = load_spacy_models()

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11166 entries, 0 to 11165
Data columns (total 45 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    11166 non-null  object 
 1   id_artist             11166 non-null  object 
 2   name_artist           11166 non-null  object 
 3   full_title            11166 non-null  object 
 4   title                 11166 non-null  object 
 5   featured_artists      3517 non-null   object 
 6   primary_artist        11166 non-null  object 
 7   language              11061 non-null  object 
 8   album                 9652 non-null   object 
 9   stats_pageviews       4642 non-null   float64
 10  swear_IT              11166 non-null  int64  
 11  swear_EN              11166 non-null  int64  
 12  swear_IT_words        11166 non-null  object 
 13  swear_EN_words        11166 non-null  object 
 14  year                  10766 non-null  object 
 15  month              

year, month, day, n_sentence, n_tokens, disc_numer, track_number, popularity should all be Int64 (or at least float64: year and popularity are object), explicit should be bool

validation helper function checking expected types validity

In [3]:
def check_type_validity(value, expected_type):
    return not isinstance(value, expected_type)

### id

In [4]:
invalid_elems = df[df['id'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['id'])

nan_indexes = df.index[df['id'].isna()].tolist()
print(f"number of missing values in id: {len(nan_indexes)}")

unique_ids = df['id'].nunique()
print(f"number of unique ids: {unique_ids}")

Series([], Name: id, dtype: object)
number of missing values in id: 0
number of unique ids: 11093


there are no missing id values in the dataset.
all the ids are strings with the following format "TR{unique_num_id}".
there are less unique ids than the number of rows (11093 < 11166), let's check if they come from duplicate songs or there are different songs with the same id

In [5]:
not_unique_ids = df['id'].value_counts()
not_unique_ids = not_unique_ids[not_unique_ids > 1]
print(f"there are {len(not_unique_ids)} not unique ids:")
print(not_unique_ids)

not_unique_ids_list = not_unique_ids.index.tolist()

identical_duplicates = []
different_duplicates = []
mixed_duplicates = [] #ids with both identical and different rows

for dup_id in not_unique_ids_list:
    duplicate_rows = df[df['id'] == dup_id]
    num_occurrences = len(duplicate_rows)
    
    unique_songs = duplicate_rows[['name_artist', 'title']].drop_duplicates()
    num_unique = len(unique_songs)

    if num_unique == 1:
        identical_duplicates.append(dup_id) #duplicate rows
    elif num_unique == num_occurrences:
        different_duplicates.append(dup_id) #different songs
    else:
        mixed_duplicates.append(dup_id) #duplicate rows and different songs

print(f"{len(identical_duplicates)} come from duplicate songs, ",
      f"{len(different_duplicates)} come from different songs with the same id, ",
      f"{len(mixed_duplicates)} have both duplicate and different songs")

there are 71 not unique ids:
id
TR367132    4
TR978886    2
TR987615    2
TR690925    2
TR772702    2
           ..
TR245683    2
TR903275    2
TR679972    2
TR247772    2
TR261964    2
Name: count, Length: 71, dtype: int64
0 come from duplicate songs,  70 come from different songs with the same id,  1 have both duplicate and different songs


of the 71 not unique ids, 70 refers to different songs, meaning that there are no duplicate songs respect to those ids, but there are ids which refers to multiple songs.

by analyzing more in depth the case with both duplicate and different songs:

In [6]:
for dup_id in mixed_duplicates: #iterating for a "general" solution, but in this case wouldn't be necessary
    duplicate_artist_song = df[df['id'] == dup_id]
    unique_songs = duplicate_artist_song[['name_artist', 'title']].drop_duplicates()
    print(f"for id {dup_id}, there are {len(unique_songs)} unique songs:")
    
    for idx, row in unique_songs.iterrows():
        song_rows = duplicate_artist_song[
            (duplicate_artist_song['name_artist'] == row['name_artist']) &
            (duplicate_artist_song['title'] == row['title'])]
        
        print(f"\t- '{row['title']}' by artist: {row['name_artist']}, occuring {len(song_rows)} times")
        
        if len(song_rows) != 1: #more than one occurrence
            first_row = song_rows.iloc[0]
            all_identical = True
            differing_columns = []

            for col in song_rows.columns:
                unique_values = song_rows[col].nunique()
                if unique_values > 1:
                    all_identical = False
                    differing_columns.append(col)
            
            if not all_identical:
                print(f"\t\tdiffering columns: {differing_columns}\n")
                
                #for col in differing_columns: #uncomment this for see values of different rows
                        #values = song_rows[col].tolist()
                        #print(f"\t\t• {col}:")
                        #for i, val in enumerate(values):
                            #print(f"\t\t\tRow {i+1}: {val}")

for id TR367132, there are 2 unique songs:
	- 'BUGIE' by artist: Madame, occuring 2 times
		differing columns: ['year', 'album_name', 'album_release_date', 'album_type', 'track_number', 'duration_ms', 'popularity', 'album_image', 'id_album']

	- '​sentimi' by artist: Madame, occuring 2 times
		differing columns: ['album_name', 'album_release_date', 'album_type', 'track_number', 'duration_ms', 'popularity', 'album_image', 'id_album']



so that id not only refers to 2 different songs, but each occurrence of the song (2) has different values. based on the values of different rows, it's possible to cut the additional wrong records

### id_artist

In [7]:
invalid_elems = df[df['id_artist'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['id_artist'])

nan_indexes = df.index[df['id_artist'].isna()].tolist()
print(f"number of missing values in id_artist: {len(nan_indexes)}")

id_to_names = df.groupby('id_artist')['name_artist'].nunique()
multiple_names = id_to_names[id_to_names > 1]

if len(multiple_names) == 0:
    print("each id_artist corresponds to exactly one name_artist")

Series([], Name: id_artist, dtype: object)
number of missing values in id_artist: 0
each id_artist corresponds to exactly one name_artist


there are no missing id_artist values in the dataset. all the id_artists are strings with the following format "ART{unique_num_id}". each id_artist corresponds to exactly one name_artist

### name_artist

In [8]:
invalid_elems = df[df['name_artist'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['name_artist'])

nan_indexes = df.index[df['name_artist'].isna()].tolist()
print(f"number of missing values in name_artist: {len(nan_indexes)}")

name_to_ids  = df.groupby('name_artist')['id_artist'].nunique()
multiple_ids  = name_to_ids [name_to_ids  > 1]

if len(multiple_names) == 0:
    print("each name_artist corresponds to exactly one id_artist")

Series([], Name: name_artist, dtype: object)
number of missing values in name_artist: 0
each name_artist corresponds to exactly one id_artist


there are no missing name_artist values in the dataset. all the name_artists are strings with the following format "ART{unique_num_id}". each name_artist corresponds to exactly one id_artist

### full_title

In [9]:
invalid_elems = df[df['full_title'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['full_title'])

nan_indexes = df.index[df['full_title'].isna()].tolist()
print(f"number of missing values in full_title: {len(nan_indexes)}")

Series([], Name: full_title, dtype: object)
number of missing values in full_title: 0


there are no missing full_title values in the dataset. 
all the full_titles are strings with the following format: "{title} by {primary_artist} (Ft. {name_artist_1}, {...}, {name_artist_n-1} & {name_artist_n})". let's check if name_artist is inside every full title

In [10]:
#normalizing non-breaking space and quotes
df['full_title'] = (df['full_title']
                    .str.replace('\xa0', ' ', regex=False)    #non-breaking space
                    .str.replace('\u2019', "'", regex=False)  #right single quote
                    .str.replace('\u2018', "'", regex=False)  #left single quote
                    .str.replace('\u201c', '"', regex=False)  #left double quote
                    .str.replace('\u201d', '"', regex=False)) #right double quote

#(.+?) -> capture everything; (?:\s*\(|$) -> stop at either " (" or end of string
extracted_artist = df['full_title'].str.extract(r' by (.+?)(?:\s*\(|$)', expand=False).str.strip()

mismatches = df[extracted_artist != df['name_artist']]
print(f"number of name_artist mismatches inside full_title: {len(mismatches)}")

if len(mismatches) > 0:
    print("sample mismatches:")
    mismatches_display = mismatches.copy()
    mismatches_display['extracted_artist'] = extracted_artist[mismatches.index]
    display(mismatches_display[['full_title', 'name_artist', 'extracted_artist']].head(20))

number of name_artist mismatches inside full_title: 844
sample mismatches:


,full_title,name_artist,extracted_artist
36,"Mattonelle* by Rosa Chemical, Mehdi (ITA), Fri...",Rosa Chemical,"Rosa Chemical, Mehdi"
129,a me mi piace by Alfa & Manu Chao,Alfa,Alfa & Manu Chao
223,EGLI È IL RE by thasup & Mara Sattei,thasup,thasup & Mara Sattei
225,POSTO MIO by thasup & Mara Sattei,thasup,thasup & Mara Sattei
227,​r()t()nda by thasup & Tiziano Ferro,thasup,thasup & Tiziano Ferro
231,SO CHE CI SEI by thasup & Mara Sattei,thasup,thasup & Mara Sattei
235,BLESS SU BLESS by thasup & Mara Sattei,thasup,thasup & Mara Sattei
238,ONE KING by thasup & Mara Sattei,thasup,thasup & Mara Sattei
241,COME POLVERE by thasup & Mara Sattei,thasup,thasup & Mara Sattei
244,OCCHI MIEI by thasup & Mara Sattei,thasup,thasup & Mara Sattei


In [11]:
multi_artist_indicators = ['&', ',', ' and ']

has_separator = df['primary_artist'].str.contains('|'.join(multi_artist_indicators), case=False, na=False)
potential_multi = df[has_separator]

print(f"records with single artist: {len(df) - len(potential_multi)}")

records with single artist: 11166


The name_artist column contains only one artist per track, even when multiple artists collaborated (excluding featured artists), likely due to a data collection system limitation. while some tracks include non-rapper collaborators not in the dataset (e.g. Tiziano Ferro, Mara Sattei), which is expected, other tracks list only one rapper despite having multiple rapper collaborators who are present in the dataset (e.g. Ensi in Cemento). this causes artist-based analyses to undercount songs, as tracks where an artist collaborated but wasn't listed as the primary artist will be excluded from their catalog.

### title

In [12]:
invalid_elems = df[df['title'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['title'])

nan_indexes = df.index[df['title'].isna()].tolist()
print(f"number of missing values in title: {len(nan_indexes)}")

Series([], Name: title, dtype: object)
number of missing values in title: 0


In [13]:
df['title'] = (df['title']
                .str.replace('\xa0', ' ', regex=False)
                .str.replace('\u2019', "'", regex=False)
                .str.replace('\u2018', "'", regex=False)
                .str.replace('\u201c', '"', regex=False)
                .str.replace('\u201d', '"', regex=False))

# ^(.+?) -> capture from start; \s+by\s+ -> stop at " by " (with flexible whitespace)
extracted_title = df['full_title'].str.extract(r'^(.+?)\s+by\s+', expand=False).str.strip()

mismatches = df[df['title'] != extracted_title]
print(f"number of title mismatches inside full_title: {len(mismatches)}")

if len(mismatches) > 0:
    print("remaining mismatches (likely titles containing 'by'):\n")
    mismatches_display = mismatches.copy()
    mismatches_display['extracted_title'] = extracted_title[mismatches.index]
    display(mismatches_display[['full_title', 'title', 'extracted_title']].head(10))

number of title mismatches inside full_title: 4
remaining mismatches (likely titles containing 'by'):



,full_title,title,extracted_title
2513,Bolo by Night RMX by Inoki,Bolo by Night RMX,Bolo
2578,Bolo by Night RMX (NewKingzTape) by Inoki (Ft....,Bolo by Night RMX (NewKingzTape),Bolo
2600,Bolo by Night (NewKingz RMX) by Inoki (Ft. Lil...,Bolo by Night (NewKingz RMX),Bolo
2642,Barona by Night (Red Bull Culture Clash dubpla...,Barona by Night (Red Bull Culture Clash dubplate),Barona


title is contained inside full_title with no errors. the missmatches appearing are titles containing the word 'by', which is what's used from the regex to split the string.

### featured_artists

In [14]:
invalid_elems = df[df['featured_artists'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['featured_artists'])

nan_indexes = df.index[df['featured_artists'].isna()].tolist()
print(f"number of missing values in featured_artists: {len(nan_indexes)}")

5        NaN
6        NaN
8        NaN
9        NaN
10       NaN
        ... 
11159    NaN
11161    NaN
11163    NaN
11164    NaN
11165    NaN
Name: featured_artists, Length: 7649, dtype: object
number of missing values in featured_artists: 7649


this is expected, since not every track must have a featured artist. let's check if the featured artists of the column are the same listed in the full_title

In [15]:
df['featured_artists'] = (df['featured_artists']
                    .str.replace('\xa0', ' ', regex=False)
                    .str.replace('\u2019', "'", regex=False)
                    .str.replace('\u2018', "'", regex=False)
                    .str.replace('\u201c', '"', regex=False)
                    .str.replace('\u201d', '"', regex=False))

# \((?:Ft\.)\s*(.+?)\)\s*$ -> everything from "(Ft.)" until the closing ")"
extracted_featured = df['full_title'].str.extract(r'\((?:Ft\.)\s*(.+?)\)\s*$', expand=False).str.strip()
extracted_featured_normalized = extracted_featured.str.replace(' & ', ', ', regex=False)

def sort_splitted_by_comma(string):
    if pd.isna(string):
        return string
    ret = [a.strip() for a in string.split(',')]
    return ', '.join(sorted(ret))

extracted_sorted = extracted_featured_normalized.apply(sort_splitted_by_comma)
featured_sorted = df['featured_artists'].apply(sort_splitted_by_comma)

mismatches = df[
    ((extracted_sorted.notna()) & (featured_sorted.isna())) |
    ((extracted_sorted.isna()) & (featured_sorted.notna())) |
    ((extracted_sorted.notna()) & (featured_sorted.notna()) & (extracted_sorted != featured_sorted))
]

print(f"Rows with featured artists in full_title: {extracted_featured.notna().sum()}")
print(f"Rows with featured artists in column: {df['featured_artists'].notna().sum()}")
print(f"number of featured_artists mismatches (ignoring order): {len(mismatches)}\n")

if len(mismatches) > 0:
    print("Sample mismatches (actual differences, not just ordering):")
    mismatches_display = mismatches.copy()
    mismatches_display['featured_sorted'] = featured_sorted[mismatches.index]
    mismatches_display['extracted_sorted'] = extracted_sorted[mismatches.index]
    display(mismatches_display[['full_title', 'featured_sorted', 'extracted_sorted']].head(10))

Rows with featured artists in full_title: 3517
Rows with featured artists in column: 3517
number of featured_artists mismatches (ignoring order): 9

Sample mismatches (actual differences, not just ordering):


,full_title,featured_sorted,extracted_sorted
1566,Comunque vada... by Piotta (Ft. Turi & Compari),Turi & Compari,"Compari, Turi"
2427,Musica Vera by Mistaman (Ft. Stokka & MadBuddy),Stokka & MadBuddy,"MadBuddy, Stokka"
2461,Il giorno in cui ci siamo svegliati by Mistama...,Stokka & MadBuddy,"MadBuddy, Stokka"
2645,Lo Facciamo Così by Inoki (Ft. Stokka & MadBuddy),Stokka & MadBuddy,"MadBuddy, Stokka"
2844,Dreams by Johnny Marsiglia & Big Joe (Ft. Stok...,Stokka & MadBuddy,"MadBuddy, Stokka"
3301,Pace by Ghemon (Ft. Franco Negrè & Stokka & Ma...,"Franco Negrè, Stokka & MadBuddy","Franco Negrè, MadBuddy, Stokka"
3616,'O Sarracino by Neffa (Ft. Giuliano Palma & Th...,Giuliano Palma & The Bluebeaters,"Giuliano Palma, The Bluebeaters"
5576,CHIAGNE by Geolier (Ft. Lazza & Takagi & Ketra),"Lazza, Takagi & Ketra","Ketra, Lazza, Takagi"
9095,DONDURÈ by Dark Polo Gang (Ft. Ski & Wok),Ski & Wok,"Ski, Wok"


featured_artists is contained inside full_title with no errors. the missmatches are given from features made with group names including the character '&', which is what's used from the regex to split the string.

### primary_artist

In [16]:
invalid_elems = df[df['primary_artist'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['primary_artist'])

nan_indexes = df.index[df['primary_artist'].isna()].tolist()
print(f"number of missing values in primary_artist: {len(nan_indexes)}")

Series([], Name: primary_artist, dtype: object)
number of missing values in primary_artist: 0


In [17]:
mismatches = df[df['primary_artist'] != df['name_artist']]
print(f"mismatches between name_artist and primary_artist: {len(mismatches)}")

if len(mismatches) > 0:
    display(mismatches[['id', 'full_title', 'name_artist', 'primary_artist']].head(5))

mismatches between name_artist and primary_artist: 0


there are no mismatches betweem this column and the name_artist column, so it doesn't add new informations. regarding the multiple primary_artists issue highlighted during the full_title analysis, it could be possible to both use this column to list all the primary artists (if more are present, and they are also present in the artist dataset), or cut this column and add more rows for the primary artists not present in name_artist but are present as artist and are present in the artist dataset.

### lyrics

In [18]:
invalid_elems = df[df['lyrics'].apply(check_type_validity, expected_type=str)]
print(invalid_elems[['full_title', 'lyrics']])

nan_indexes = df.index[df['lyrics'].isna()].tolist()
print(f"number of missing values in lyrics: {len(nan_indexes)}")

                           full_title lyrics
6950   Consegnato.02 by Frah Quintale    NaN
6952   Consegnato.01 by Frah Quintale    NaN
9775  Introliberassociazione by Salmo    NaN
number of missing values in lyrics: 3


between the 3 trakcs with no lyrics, only 'Consegnato.01' is the one who actually is missing some lyrics, the other ones are instrumental tracks. While those two instrumental tracks have lyrics set to NaN, there are other instrumenal/base tracks with the following form "{n} Contributor/s{title}". In those tracks some columns such as language, n_tokens, n_sentences, tokens_per_sent, char_per_tok, avg_token_per_clause, lexical_density and swear columns would be meaningless, so the lyrics will be set to NaN and those values will be set to NaN/0.

A few of those records contain a description of the track, those will be set to NaN. 13 other tracks instead contain the lyrics in the following form "{n} Contributor(s){title} Lyrics{lyrics}". Those lyrics will be kept, but most of them are missing new lines, so they are stored in single lines.

In [19]:
#list of tracks with lyrics inside
rescue_list = [
    "Intro (Napolimanicomio)",
    "Dammi Ancora",
    "Cin Cin",
    "Tu Non Hai Mai",
    "Pronte A Tutto",
    "Non Dimentico Più",
    "Come Ti Senti?",
    "0 tempo e 0 vento",
    "Vita Vera - Story",
    "Meteoriti (english)",
    "24h Non Stop",
    "Dimmi Che Farai",
    "Se ti girassi"
]

#regex to find the "Lyrics" part and capture everything after
lyrics_pattern = re.compile(r'^\d+\s+Contributor(s)?.*?Lyrics(.*)', re.DOTALL)

def clean_contributor_lyrics(row):
    if pd.isna(row['lyrics']):
        return np.nan
        
    if row['title'] in rescue_list:
        match = lyrics_pattern.search(row['lyrics'])
        if match:
            rescued_text = match.group(2).strip()
            
            if not rescued_text: #if text after regex is empty
                return np.nan
            else:
                return rescued_text
        else:
            print(f"regex failed for target track: {row['title']}")
            return row['lyrics']
    else:
        #not a target track
        return row['lyrics']
    
target_mask = df['title'].isin(rescue_list)

lyrics_before = df.loc[target_mask, 'lyrics'].copy()
df['lyrics'] = df.apply(clean_contributor_lyrics, axis=1)
lyrics_after = df.loc[target_mask, 'lyrics']

rescued_count = (lyrics_before != lyrics_after).sum()

print(f"successfully rescued lyrics for {rescued_count} out of {len(rescue_list)} tracks.")

regex failed for target track: Cin Cin
successfully rescued lyrics for 13 out of 13 tracks.


the regex failed comes from another track named Cin Cin made from a different artist with correctly formatted lyrics, but the 13 tracks with the lyrics were mantained.

Now the others bad lyrics tracks will be nullified, along with meaningless fields without the lyrics:

In [20]:
def nullify_bad_lyrics(lyrics): #sets lyrics for instrumentals, metadata, and descriptions to NaN.
    if pd.isna(lyrics):
        return np.nan
    
    #nullify remaining "contributor"
    if 'Contributor' in lyrics:
        return np.nan
    
    return lyrics

nans_before = df['lyrics'].isna().sum()
df['lyrics'] = df['lyrics'].apply(nullify_bad_lyrics)
nans_after = df['lyrics'].isna().sum()

newly_nullified_count = nans_after - nans_before
print(f"nullified {newly_nullified_count} additional rows.")

nullified 125 additional rows.


In [21]:
#columns that are meaningless without lyrics
count_cols = [
    'n_tokens', 'n_sentences', 'swear_IT', 'swear_EN'
]

ratio_cols = [
    'tokens_per_sent', 'char_per_tok', 
    'avg_token_per_clause', 'lexical_density'
]

string_cols = [
    'language', 'swear_IT_words', 'swear_EN_words'
]

missing_lyrics_mask = df['lyrics'].isna()

df.loc[missing_lyrics_mask, count_cols] = 0
df.loc[missing_lyrics_mask, ratio_cols] = 0

# Use pd.NA for string/object columns
for col in string_cols:
    df.loc[missing_lyrics_mask, col] = pd.NA

print(f"Cleaned {missing_lyrics_mask.sum()} rows with missing lyrics.")

Cleaned 128 rows with missing lyrics.


### language

In [22]:
invalid_elems = df[df['language'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['language'])

nan_indexes = df.index[df['language'].isna()].tolist()
print(f"number of missing values in language: {len(nan_indexes)}")

58       <NA>
61       <NA>
65       <NA>
68       <NA>
198       NaN
         ... 
10334    <NA>
10649    <NA>
10653    <NA>
10684    <NA>
10691    <NA>
Name: language, Length: 203, dtype: object
number of missing values in language: 203


In [23]:
language_counts = df['language'].value_counts(dropna=False)

for i, (lang, count) in enumerate(language_counts.items(), 1):
    percentage = (count / len(df)) * 100
    lang_display = lang if pd.notna(lang) else "NaN"
    print(f"{i:3d}. {lang_display:13s} : {count:5d} tracks")

  1. it            :  9288 tracks
  2. en            :   814 tracks
  3. pl            :   585 tracks
  4. NaN           :   128 tracks
  5. NaN           :    75 tracks
  6. es            :    63 tracks
  7. co            :    48 tracks
  8. pt            :    45 tracks
  9. da            :    18 tracks
 10. cs            :    14 tracks
 11. fr            :    13 tracks
 12. nl            :     9 tracks
 13. war           :     9 tracks
 14. sr            :     9 tracks
 15. eu            :     8 tracks
 16. no            :     6 tracks
 17. ca            :     5 tracks
 18. ia            :     5 tracks
 19. gl            :     3 tracks
 20. rm            :     2 tracks
 21. la            :     2 tracks
 22. de            :     2 tracks
 23. sco           :     2 tracks
 24. eo            :     2 tracks
 25. et            :     1 tracks
 26. lt            :     1 tracks
 27. rw            :     1 tracks
 28. aa            :     1 tracks
 29. ro            :     1 tracks
 30. ru       

here are all the languages in the dataset. a lot of entries here are incorrect, and the criteria by which the language was assigned is not known. the following is a list of language codes that, upon manual inspection, actually correspond to Italian tracks: pl (Polish), da (Danish), cs (Czech), nl (Dutch), sr (Serbian), war (?), eu (?), no (?), ia (Interlingua), ca (Catalan), gl (Galician), sco (?), la (Latin), eo (Esperanto), rm (Romansh), et (Estonian), lt (Lithuanian), aa (?), ro (Romanian), rw (Kinyarwanda), chr (?), qu (Quechua), mt (Maltese), cy (Welsh), sq (Albanian), sw (Swahili).

The first NaN are the Pandas' NA that were introduced in the previous step for tracks with no lyrics, while the 75 others are original missing values from the dataset, which will try to be filled with langdetect.

In [24]:
# NaN language detection with langdetect
def detect_language(lyrics):
    #handle NaN, empty, or non-string lyrics
    if pd.isna(lyrics) or not isinstance(lyrics, str) or not lyrics.strip():
        return pd.NA

    try:
        return detect(lyrics)
    except LangDetectException:
        #handle cases with no detectable features (e.g., "!!!", "123")
        return pd.NA
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        return pd.NA

#language missing and lyrics not missing
mask_to_check = df['language'].isna() & df['lyrics'].notna()
    
rows_to_check = df[mask_to_check]
print(f"Found {len(rows_to_check)} rows with missing language but existing lyrics")

if not rows_to_check.empty:
    detected_languages = rows_to_check['lyrics'].apply(detect_language)
    df.loc[detected_languages.index, 'language'] = detected_languages

# fix incorrect italian codes
incorrect_italian_codes = ['pl', 'da', 'cs', 'nl', 'sr', 'war', 'eu', 'no', 'ia', 'ca', 'gl', 'sco', 'la', 'eo', 'rm', 'et', 'lt', 'aa', 'ro', 'rw', 'chr', 'qu', 'mt', 'cy', 'sq', 'sw', 'co']
updated_rows = df[df['language'].isin(incorrect_italian_codes)]
print(f"updated rows: {len(updated_rows)}")

df.loc[df['language'].isin(incorrect_italian_codes), 'language'] = 'it'

language_counts = df['language'].value_counts(dropna=False)
for i, (lang, count) in enumerate(language_counts.items(), 1):
    percentage = (count / len(df)) * 100
    lang_display = lang if pd.notna(lang) else "NaN (missing)"
    print(f"{i:3d}. {lang_display:13s} : {count:5d} tracks")

Found 75 rows with missing language but existing lyrics
updated rows: 737
  1. it            : 10092 tracks
  2. en            :   818 tracks
  3. NaN (missing) :   128 tracks
  4. es            :    64 tracks
  5. pt            :    47 tracks
  6. fr            :    13 tracks
  7. de            :     2 tracks
  8. ru            :     1 tracks
  9. bg            :     1 tracks


some (en (English), es (Spanish), pt (Portuguese), fr (French)) are mixed, co (Corsican) is basically italian (merged with 'it' as well), while the rest (de (German), ru (Russian), bg (Bulgarian)) are good. langdetect will be used for the mixed columns to obtain better results.

Ref. for languages code: https://www.cambridge.org/core/books/abs/interaction-of-borrowing-and-word-formation/list-of-iso639-language-codes/624C593339914D4A7DDA1AC55AC4889F

In [25]:
mixed_languages = ['en', 'es', 'pt', 'fr']
mask_to_check = df['language'].isin(mixed_languages) & df['lyrics'].notna()
rows_to_check = df[mask_to_check]

if not rows_to_check.empty:
    detected_languages = rows_to_check['lyrics'].apply(detect_language)

    #detected langugage distribution
    new_counts = detected_languages.value_counts(dropna=False)
    print(new_counts)

    target_langs = ['sl', 'ca', 'nl', 'so']
    target_indices = detected_languages[detected_languages.isin(target_langs)].index
    if not target_indices.empty:
        print(f"target languages to manually check ({target_langs}):")
        
        target_tracks = df.loc[target_indices].copy()
        
        target_tracks['detected_language'] = detected_languages[target_indices]
        target_tracks['lyrics_snippet'] = target_tracks['lyrics'].str[:100] + "..."
        
        display(target_tracks[[
            'title', 
            'language',          
            'detected_language',
            'lyrics_snippet'
        ]])
    else:
        print(f"no tracks found matching {target_langs}.")
    
    #update rows where detection was successful
    update_mask = detected_languages.notna()
    rows_to_update = detected_languages[update_mask]

    if not rows_to_update.empty:
        print(f"updating {len(rows_to_update)} rows in the 'language' column")
        
        df.loc[rows_to_update.index, 'language'] = rows_to_update.values

else:
    print("No rows found to check.")

lyrics
it    745
en    143
es     39
pt      6
fr      5
sl      1
ca      1
nl      1
so      1
Name: count, dtype: int64
target languages to manually check (['sl', 'ca', 'nl', 'so']):


,title,language,detected_language,lyrics_snippet
2802,KI-KI,en,sl,"Soy yo, la Ferrari (Dímelo)\nAcelera color Cam..."
3479,The Banana Splits,en,ca,"Tra-la-la, la-la-la-la\nTra-la-la, la-la-la-la..."
4373,Intro (Monkee Bizniz Vol. 2),en,nl,Monkee Bizniz Vol. 2\n Busdeez\nYou know my na...
8533,FOXY,en,so,"Sex (Oh, ah, ah, ah, oh, ah)\nDrugs (Oh, ah, a..."


updating 942 rows in the 'language' column


in KI-KI the main language is spanish, in The Banana Splits, Intro (Monkee Bizniz Vol. 2), and FOXY the language was correctly english, so they will be updated accordingly

In [26]:
manual_updates = {
    "KI-KI": "es",
    "The Banana Splits": "en",
    "Intro (Monkee Bizniz Vol. 2)": "en",
    "FOXY": "en"
}

updated_count = 0

for title, lang in manual_updates.items():
    mask = (df['title'] == title)
    
    # Check if any rows were found
    if mask.any():
        df.loc[mask, 'language'] = lang
        updated_count += mask.sum()
    else:
        print(f"track with title '{title}' not found in dataset.")
        
print(f"rows updated: {updated_count}")

print(f"\nFinal language distribution:")
language_counts = df['language'].value_counts(dropna=False)

for i, (lang, count) in enumerate(language_counts.items(), 1):
    percentage = (count / len(df)) * 100
    lang_display = lang if pd.notna(lang) else "NaN (missing)"
    print(f"{i:3d}. {lang_display:13s} : {count:5d} tracks")

rows updated: 4

Final language distribution:
  1. it            : 10837 tracks
  2. en            :   146 tracks
  3. NaN (missing) :   128 tracks
  4. es            :    40 tracks
  5. pt            :     6 tracks
  6. fr            :     5 tracks
  7. de            :     2 tracks
  8. ru            :     1 tracks
  9. bg            :     1 tracks


### album

In [27]:
invalid_elems = df[df['album'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['album'])

nan_indexes = df.index[df['album'].isna()].tolist()
print(f"number of missing values in album: {len(nan_indexes)}")

10       NaN
21       NaN
22       NaN
23       NaN
24       NaN
        ... 
11156    NaN
11157    NaN
11159    NaN
11164    NaN
11165    NaN
Name: album, Length: 1514, dtype: object
number of missing values in album: 1514


this is expected, since not every track must be inside an album.

In [28]:
tracks_per_album = df['album'].value_counts()

print('top 20 albums with most tracks:')
for i, (album, count) in enumerate(tracks_per_album.head(20).items(), 1):
    print(f"{i:3d}. {album[:50]:50s} : {count:3d} tracks")

top 20 albums with most tracks:
  1. Rimo Da Quando                                     :  47 tracks
  2. TAXI DRIVER (Deluxe)                               :  37 tracks
  3. Solo Un Uomo (Gold Edition)                        :  36 tracks
  4. Il tempo vola 2002-2020                            :  36 tracks
  5. Di vizi di forma virtù                             :  35 tracks
  6. Cose Dell’Altro Mondo Mondo Marcio: Special Editio :  34 tracks
  7. SurreAle                                           :  30 tracks
  8. Status (Vendetta Edition)                          :  30 tracks
  9. DIO LO SA - ATTO II                                :  30 tracks
 10. Re Mida: Aurum                                     :  29 tracks
 11. The NewKingzTape Vol. 1                            :  29 tracks
 12. Miracolo! (Deluxe Edition)                         :  28 tracks
 13. ’A Verità 2.0                                      :  27 tracks
 14. Una cosetta così                                   :  26 tracks
 1

some albums list more tracks than they actually contain. my best guess is that a few standalone songs were mistakenly assigned to one of the artist’s albums.

### stats_pageviews

In [29]:
invalid_elems = df[df['stats_pageviews'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['stats_pageviews'])

nan_indexes = df.index[df['stats_pageviews'].isna()].tolist()
print(f"number of missing values in stats_pageviews: {len(nan_indexes)}")

0        196033.0
1        164450.0
2         58313.0
3         39890.0
4         35432.0
           ...   
11161         NaN
11162         NaN
11163         NaN
11164         NaN
11165         NaN
Name: stats_pageviews, Length: 11166, dtype: float64
number of missing values in stats_pageviews: 6524


In [30]:
df['stats_pageviews'] = df['stats_pageviews'].astype('Int64')

invalid_elems = df[df['stats_pageviews'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['stats_pageviews'])

nan_indexes = df.index[df['stats_pageviews'].isna()].tolist()
print(f"number of missing values in stats_pageviews: {len(nan_indexes)}")

0        196033
1        164450
2         58313
3         39890
4         35432
          ...  
11161      <NA>
11162      <NA>
11163      <NA>
11164      <NA>
11165      <NA>
Name: stats_pageviews, Length: 11166, dtype: Int64
number of missing values in stats_pageviews: 6524


converted stats_pageviews to Int64 (nullable integer, NaN shown as <NA>), since the values in the column are without a decimal point.

In [31]:
invalid_pageviews = df[(df['stats_pageviews'] < 0)]
if len(invalid_pageviews) > 0:
    print(f"{len(invalid_pageviews)} rows with invalid pageviews (negative values):")
    display(invalid_pageviews[['id', 'title', 'name_artist', 'stats_pageviews']].head(10))

chart_data = df[['stats_pageviews']].dropna().copy()
chart_data['stats_pageviews'] = pd.to_numeric(chart_data['stats_pageviews']) #float is safer for Altair

base = alt.Chart(chart_data).encode(
    x=alt.X('stats_pageviews', bin=alt.Bin(maxbins=30), title='Pageviews Range'),
    y=alt.Y('count()', title='Number of Tracks')
).properties(
    title='Distribution of Tracks by Pageviews Range',
    width=1000,
    height=400
)

bars = base.mark_bar(color='steelblue', opacity=0.7)
# text labels on top of the bars
text = base.mark_text(
    dy=-5,
    color='black'
).encode(
    text=alt.Text('count()')
)

chart = (bars + text).interactive()

display(chart)

alt.LayerChart(...)

there seems to be outliers in the distribution, let's analyze the top pageviews tracks

In [32]:
top_10_tracks = df.sort_values(by='stats_pageviews', ascending=False).head(10)

print("top 10 tracks by pageviews:")
display(top_10_tracks[['title', 'name_artist', 'stats_pageviews']])

top 10 tracks by pageviews:


,title,name_artist,stats_pageviews
8188,Veleno 6,MadMan,1971207
7028,Veleno 7,Gemitaiz,762450
3046,Cupido,Sfera Ebbasta,692900
3047,Tran Tran,Sfera Ebbasta,499980
1197,Giovane fuoriclasse,Capo Plaza,484090
8189,Bolla papale Freestyle,MadMan,477926
3048,Sciroppo,Sfera Ebbasta,456566
3740,La legge del più forte,Tedua,448993
198,​blun7 a swishland,thasup,448797
4374,Ninna Nanna,Ghali,423119


all the top stats_pageviews tracks are popular ones, so those high values seems to be realistic. these values probably indicate the number of track page views on the original website where the dataset was obtained.

### swear_IT_words

In [33]:
invalid_elems = df[df['swear_IT_words'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['swear_IT_words'])

nan_indexes = df.index[df['swear_IT_words'].isna()].tolist()
print(f"number of missing values in swear_IT_words: {len(nan_indexes)}")

58       <NA>
61       <NA>
65       <NA>
68       <NA>
264      <NA>
         ... 
10334    <NA>
10649    <NA>
10653    <NA>
10684    <NA>
10691    <NA>
Name: swear_IT_words, Length: 128, dtype: object
number of missing values in swear_IT_words: 128


In [34]:
import ast

all_it_swear_words = set()

for value in df['swear_IT_words']:
    if pd.notna(value) and value != '[]':  #skip NaN and empty lists
        try:
            word_list = ast.literal_eval(value)
            all_it_swear_words.update(word_list)
        except:
            pass
        
print(all_it_swear_words)

{'coglione', 'cretino', 'idiozia', 'cazzeggiare', 'ricchione', 'merdina', 'vacca', 'peluria', 'incazzato', 'pisello', 'mignotta', 'gay', 'bernarda', 'bombare', 'pugnetta', 'seccatore', 'vagina', 'fottersi', 'culo', 'sputtanare', 'granchio', 'incazzarsi', 'vaccata', 'scassare', 'merda', 'fottere', 'segaiolo', 'fesso', 'cagata', 'merdata', 'mazzo', 'fica', 'pompino', 'pisciata', 'sborra', 'cazzata', 'bagascia', 'jug', 'cappella', 'nerchia', 'piscio', 'coglioni', 'cagare', 'paraculo', 'uccello', 'incazzare', 'cornuto', 'rompicoglioni', 'fava', 'sborrare', 'checca', 'escremento', 'pecorina', 'bocchinaro', 'schizzare', 'toro', 'water', 'sfiga', 'puttanata', 'cesso', 'sedere', 'chiavata', 'scoreggia', 'sorca', 'sgualdrina', 'coglionata', 'scazzato', 'figa', 'fregarsene', 'sveltina', 'fogne', 'gigolo', 'cacata', 'pippa', 'minchiata', 'cazzo', 'cretini', 'mezzasega', 'stronzo', 'spompinare', 'handicappato', 'cretina', 'leccaculo', 'scrofa', 'fottuto', 'inculare', 'farabutto', 'grilletto', 'der

there seems to be some english words among the italian swear words: "water, blowjob, jug"; "gay" is also english, but it doesn't appean in the swear_EN_words, so it was probably counted as an italian word. also there are words that are not swear words ("toro, fortuna, zanzara, granchio"), and other that are 'neutral' words ("pene, vagina, glutei, sedere, escremento, feci, peluria, selvaggio") but can sound vulgar in certain tracks. note that not all the swear words are taken into consideration, since some of them inside the lyrics can be a bit 'creative' (e.g. "apputtanamento")

In [35]:
it_words_to_exclude = ['water', 'blowjob', 'jug', 'toro', 'fortuna', 'zanzara', 'granchio']
filtered_it_swear_words = {word for word in all_it_swear_words if word not in it_words_to_exclude}

In [36]:
def find_swear_words_in_lyrics(lyrics, swear_words_list):
    if pd.isna(lyrics):
        return []

    lyrics_lower = lyrics.lower() #case-insensitive matching
    
    found_words = []
    for word in swear_words_list:
        pattern = r'\b' + re.escape(word) + r'\b'
        if re.search(pattern, lyrics_lower):
            found_words.append(word)
    
    return found_words
            
rows_changed = 0
rows_increased = 0
rows_decreased = 0
rows_both_changed = 0
rows_unchanged = 0
rows_processed = 0

for idx in df.index:
    lyrics = df.loc[idx, 'lyrics']
    
    found_words = find_swear_words_in_lyrics(lyrics, filtered_it_swear_words)
    new_words = set(found_words)
    
    old_value = df.loc[idx, 'swear_IT_words']
    
    #parse old swear words
    try:
        if pd.isna(old_value) or old_value == '[]':
            old_words = set()
        else:
            old_words = set(ast.literal_eval(old_value))
    except:
        old_words = set()
    
    new_value = str(found_words)
    df.loc[idx, 'swear_IT_words'] = new_value
    
    #track changes
    if old_words != new_words:
        rows_changed += 1
        
        words_added = new_words - old_words
        words_removed = old_words - new_words
        
        if words_added and words_removed:
            rows_both_changed += 1
        elif words_added:
            rows_increased += 1
        elif words_removed:
            rows_decreased += 1
    else:
        rows_unchanged += 1
    
    rows_processed += 1

print(f"unchanged rows: {rows_processed - rows_changed}")
print(f"rows where swear_IT_words was changed: {rows_changed}")

print(f"\t- words added only: {rows_increased}")
print(f"\t- words removed only: {rows_decreased}")
print(f"\t- both added and removed: {rows_both_changed}")

unchanged rows: 10659
rows where swear_IT_words was changed: 507
	- words added only: 0
	- words removed only: 507
	- both added and removed: 0


### swear_EN_words

In [37]:
invalid_elems = df[df['swear_EN_words'].apply(check_type_validity, expected_type=str)]
print(invalid_elems['swear_EN_words'])

nan_indexes = df.index[df['swear_EN_words'].isna()].tolist()
print(f"number of missing values in swear_EN_words: {len(nan_indexes)}")

58       <NA>
61       <NA>
65       <NA>
68       <NA>
264      <NA>
         ... 
10334    <NA>
10649    <NA>
10653    <NA>
10684    <NA>
10691    <NA>
Name: swear_EN_words, Length: 128, dtype: object
number of missing values in swear_EN_words: 128


In [38]:
import ast

all_en_swear_words = set()

for value in df['swear_EN_words']:
    if pd.notna(value) and value != '[]':  #skip NaN and empty lists
        try:
            word_list = ast.literal_eval(value)
            all_en_swear_words.update(word_list)
        except:
            pass
        
print(all_en_swear_words)

{'anus', 'suck', 'deepthroat', 'dildo', 'nude', 'sex', 'vagina', 'escort', 'fuckin', 'kinky', 'neonazi', 'semen', 'panties', 'milf', 'shit', 'snatch', 'sexo', 'cumshot', 'raping', 'rimming', 'cialis', 'cock', 'slut', 'playboy', 'xxx', 'voyeur', 'poof', 'hardcore', 'dick', 'sucks', 'busty', 'nigga', 'asshole', 'pedobear', 'fuck', 'ecchi', 'anal', 'spic', 'tit', 'nympho', 'butt', 'cum', 'fag', 'punany', 'threesome', 'viagra', 'creampie', 'fisting', 'topless', 'clit', 'bbw', 'scat', 'bastardo', 'shibari', 'tranny', 'porno', 'boobs', 'fellatio', 'bukkake', 'masturbation', 'lolita', 'bitches', 'skeet', 'fucking', 'faggot', 'blowjob', 'xx', 'domination', 'horny', 'cunt', 'porn', 'rape', 'vulva', 'bondage', 'motherfucker', 'rapist', 'pussy', 'bastard', 'bitch', 'bullshit', 'sexual', 'negro', 'cumming', 'gangbang', 'nipple', 'tits', 'hentai', 'titty', 'doggystyle', 'coon', 'sexy', 'pissing', 'ass', 'shitty', 'cocks', 'hooker'}


there's an italian word "bastardo" among the english swear words. as in the italian set of swear words, there are words that aren't necessary swear words, but in certain tracks can be perceived as vulgar.

In [39]:
en_words_to_exclude = ['bastardo']
filtered_en_swear_words = {word for word in all_en_swear_words if word not in en_words_to_exclude}

In [40]:
rows_changed = 0
rows_increased = 0
rows_decreased = 0
rows_both_changed = 0
rows_unchanged = 0
rows_processed = 0

for idx in df.index:
    lyrics = df.loc[idx, 'lyrics']
    
    found_words = find_swear_words_in_lyrics(lyrics, filtered_en_swear_words)
    new_words = set(found_words)
    
    old_value = df.loc[idx, 'swear_EN_words']
    
    #parse old swear words
    try:
        if pd.isna(old_value) or old_value == '[]':
            old_words = set()
        else:
            old_words = set(ast.literal_eval(old_value))
    except:
        old_words = set()
    
    new_value = str(found_words)
    df.loc[idx, 'swear_EN_words'] = new_value
    
    #track changes
    if old_words != new_words:
        rows_changed += 1
        
        words_added = new_words - old_words
        words_removed = old_words - new_words
        
        if words_added and words_removed:
            rows_both_changed += 1
        elif words_added:
            rows_increased += 1
        elif words_removed:
            rows_decreased += 1
    else:
        rows_unchanged += 1
    
    rows_processed += 1

print(f"unchanged rows: {rows_processed - rows_changed}")
print(f"rows where swear_EN_words was changed: {rows_changed}")

print(f"\t- words added only: {rows_increased}")
print(f"\t- words removed only: {rows_decreased}")
print(f"\t- both added and removed: {rows_both_changed}")

unchanged rows: 10954
rows where swear_EN_words was changed: 212
	- words added only: 0
	- words removed only: 212
	- both added and removed: 0


### swear_IT

In [41]:
invalid_elems = df[df['swear_IT'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['swear_IT'])

nan_indexes = df.index[df['swear_IT'].isna()].tolist()
print(f"number of missing values in swear_IT: {len(nan_indexes)}")

Series([], Name: swear_IT, dtype: int64)
number of missing values in swear_IT: 0


In [42]:
def count_swear_words_in_lyrics(lyrics, swear_words_list):
    if pd.isna(lyrics):
        return 0
    
    lyrics_lower = lyrics.lower()
    total_count = 0
    
    for word in swear_words_list:
        pattern = r'\b' + re.escape(word) + r'\b'
        matches = re.findall(pattern, lyrics_lower)
        total_count += len(matches)
    
    return total_count

rows_processed = 0
rows_changed = 0

for idx in df.index:
    lyrics = df.loc[idx, 'lyrics']
    old_count = df.loc[idx, 'swear_IT']
    
    swear_words_value = df.loc[idx, 'swear_IT_words']
    
    try:
        if pd.isna(swear_words_value) or swear_words_value == '[]':
            words_to_count = []
        else:
            words_to_count = ast.literal_eval(swear_words_value)
    except:
        words_to_count = []
    
    new_count = count_swear_words_in_lyrics(lyrics, words_to_count)
    
    df.loc[idx, 'swear_IT'] = new_count
    
    if old_count != new_count:
        rows_changed += 1
    
    rows_processed += 1
    
print(f"unchanged rows: {rows_processed - rows_changed}")
print(f"rows where swear_IT_words was updated: {rows_changed}")

unchanged rows: 10659
rows where swear_IT_words was updated: 507


### swear_EN

In [43]:
invalid_elems = df[df['swear_EN'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['swear_EN'])

nan_indexes = df.index[df['swear_EN'].isna()].tolist()
print(f"number of missing values in swear_EN: {len(nan_indexes)}")

Series([], Name: swear_EN, dtype: int64)
number of missing values in swear_EN: 0


In [44]:
rows_processed = 0
rows_changed = 0

for idx in df.index:
    lyrics = df.loc[idx, 'lyrics']
    old_count = df.loc[idx, 'swear_EN']
    
    swear_words_value = df.loc[idx, 'swear_EN_words']
    
    try:
        if pd.isna(swear_words_value) or swear_words_value == '[]':
            words_to_count = []
        else:
            words_to_count = ast.literal_eval(swear_words_value)
    except:
        words_to_count = []
    
    new_count = count_swear_words_in_lyrics(lyrics, words_to_count)
    
    df.loc[idx, 'swear_EN'] = new_count
    
    if old_count != new_count:
        rows_changed += 1
    
    rows_processed += 1
    
print(f"unchanged rows: {rows_processed - rows_changed}")
print(f"rows where swear_EN_words was updated: {rows_changed}")

unchanged rows: 10954
rows where swear_EN_words was updated: 212


### year

In [45]:
invalid_elems = df[df['year'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['year'])

nan_indexes = df.index[df['year'].isna()].tolist()
print(f"number of missing values in year: {len(nan_indexes)}")

0        2021.0
1        2020.0
2        2021.0
3        2019.0
4        2020.0
          ...  
11161       NaN
11162    2015.0
11163       NaN
11164    2005.0
11165       NaN
Name: year, Length: 11166, dtype: object
number of missing values in year: 400


In [46]:
non_numeric = df[pd.to_numeric(df['year'], errors='coerce').isna() & df['year'].notna()]
print(f"non-numeric year values ({len(non_numeric)}):")
print(non_numeric['year'].unique())

df['year'] = pd.to_numeric(df['year'], errors='coerce').astype('Int64')

invalid_elems = df[df['year'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['year'])

nan_indexes = df.index[df['year'].isna()].tolist()
print(f"number of missing values in year: {len(nan_indexes)}")

non-numeric year values (38):
['nPmvFS' 'MoKKcY' 'PdLLvs' 'Tr5ghq' 'FZM9N5' '3KBIhx' '7arrc9' '0S4bYq'
 'Ww2RM0' 'YDWvPA' 'v9Gs4k' '1uWHmg' 's7ujOB' 'CeZdpY' 'DkWnRx' 'oiiWuy'
 'IlHfzv' 'uUsYUb' 'unxIUH' 'iW9Q9J' 'td0lg8' 'fNBiFf' 'RqnJ7g' 'VwVFFX'
 'Ise7dm' 'vgGgkT' 'QPNIJj' 'BMMnal' 'k2QnmQ' '6R1q9L' 'P03Hkq' 'tMcdp6'
 'epqCR2' 'Tl2Dgz' 'uiux8t' 'gZ2JDM' 'Gv70wC' '8i0piY']
0        2021
1        2020
2        2021
3        2019
4        2020
         ... 
11161    <NA>
11162    2015
11163    <NA>
11164    2005
11165    <NA>
Name: year, Length: 11166, dtype: Int64
number of missing values in year: 438


In [47]:
invalid_years = df[(df['year'] < 1973) | (df['year'] > 2025)]

if len(invalid_years) > 0:
    print(f"found {len(invalid_years)} rows with invalid years (< 1973 or > 2025):")
    display(invalid_years[['id', 'title', 'name_artist', 'year']].head(10))
    
year_counts = df['year'].value_counts().sort_index()

year_counts_df = year_counts.reset_index()
year_counts_df.columns = ['Year', 'Number of Tracks']

chart = alt.Chart(year_counts_df).mark_bar(
    color='steelblue',
    opacity=0.7
).encode(
    x=alt.X('Year:Q', 
            axis=alt.Axis(format='d', labelAngle=-45), 
            title='Year'),
    y=alt.Y('Number of Tracks', title='Number of Tracks')
).properties(
    title='Distribution of Tracks per Year',
    width=1000,
    height=400
).interactive()

display(chart)

found 2153 rows with invalid years (< 1973 or > 2025):


,id,title,name_artist,year
6,TR170793,LOBBY WAY,Rosa Chemical,2072
7,TR627195,SLATT,Rosa Chemical,2098
22,TR634516,Freestyle #redbullnext #essemagazine,Rosa Chemical,1917
29,TR242610,Sono Io !!,Rosa Chemical,1920
34,TR344681,ALIENO,Rosa Chemical,1901
35,TR945862,Teschio,Rosa Chemical,2052
48,TR225367,TIPO DI N,Rosa Chemical,2081
53,TR904495,Dovreicomprarmiunacollanacomeimigos,Rosa Chemical,2060
58,TR440692,LATTE +,Rosa Chemical,1933
59,TR141416,I need some sleep (Intro),Rosa Chemical,1928


alt.Chart(...)

there seems to be a lot of invalid years. the ones above the year 2025 are for obvious reasons wrong, but the ones below 1973 might be really old tracks. let's check this by using the lowest active_start field of the artists.csv dataset

In [48]:
artists_path = path.join('..', 'dataset', 'artists.csv')
df_artists = pd.read_csv(artists_path, sep=';')

df_artists['active_start'] = pd.to_datetime(df_artists['active_start'], errors='coerce')
missing_count = df_artists['active_start'].isna().sum()
print(f"missing active_start values: {missing_count}")

earliest_date = df_artists['active_start'].min()
print(f"earliest active_start: {earliest_date}")

missing active_start values: 54
earliest active_start: 1988-01-01 00:00:00


by checking them, we can notice that are there are no artist in the dataset with such low year of active_start. there might be one in the missing active_start values that is lower, but by checking some of the tracks with release year below 1973, i can say with enough confidence that those are all invalid years.

### month

In [49]:
invalid_elems = df[df['month'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['month'])

nan_indexes = df.index[df['month'].isna()].tolist()
print(f"number of missing values in month: {len(nan_indexes)}")

0         4.0
1         3.0
2         2.0
3         3.0
4         5.0
         ... 
11161     NaN
11162    11.0
11163     NaN
11164     1.0
11165     NaN
Name: month, Length: 11166, dtype: float64
number of missing values in month: 1197


In [50]:
df['month'] = pd.to_numeric(df['month'], errors='coerce').astype('Int64')

invalid_elems = df[df['month'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['month'])

nan_indexes = df.index[df['month'].isna()].tolist()
print(f"number of missing values in month: {len(nan_indexes)}")

0           4
1           3
2           2
3           3
4           5
         ... 
11161    <NA>
11162      11
11163    <NA>
11164       1
11165    <NA>
Name: month, Length: 11166, dtype: Int64
number of missing values in month: 1197


In [51]:
invalid_months = df[(df['month'] < 1) | (df['month'] > 12)]

if len(invalid_months) > 0:
    print(f"found {len(invalid_years)} rows with invalid month (< 1 or > 12):")
    display(invalid_months[['id', 'title', 'name_artist', 'month']].head(10))

even though all the months have valid values, a lot of them are missing

### day

In [52]:
invalid_elems = df[df['day'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['day'])

nan_indexes = df.index[df['day'].isna()].tolist()
print(f"number of missing values in day: {len(nan_indexes)}")

0         2.0
1         6.0
2        19.0
3         8.0
4        29.0
         ... 
11161     NaN
11162     9.0
11163     NaN
11164     1.0
11165     NaN
Name: day, Length: 11166, dtype: float64
number of missing values in day: 1323


In [53]:
df['day'] = pd.to_numeric(df['day'], errors='coerce').astype('Int64')

invalid_elems = df[df['day'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['day'])

nan_indexes = df.index[df['day'].isna()].tolist()
print(f"number of missing values in day: {len(nan_indexes)}")

0           2
1           6
2          19
3           8
4          29
         ... 
11161    <NA>
11162       9
11163    <NA>
11164       1
11165    <NA>
Name: day, Length: 11166, dtype: Int64
number of missing values in day: 1323


### n_sentences

In [54]:
invalid_elems = df[df['n_sentences'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['n_sentences'])

nan_indexes = df.index[df['n_sentences'].isna()].tolist()
print(f"number of missing values in n_sentences: {len(nan_indexes)}")

0        102.0
1         56.0
2         88.0
3         37.0
4         48.0
         ...  
11161     75.0
11162     76.0
11163     74.0
11164     32.0
11165     45.0
Name: n_sentences, Length: 11166, dtype: float64
number of missing values in n_sentences: 70


In [55]:
df['n_sentences'] = pd.to_numeric(df['n_sentences'], errors='coerce').astype('Int64')

def count_sentences_in_lyrics(lyrics):
    if pd.isna(lyrics):
        return 0
    lines = lyrics.split('\n')
    count = sum(1 for line in lines if line.strip())
    return count

def count_empty_lines(lyrics):
    if pd.isna(lyrics):
        return 0
    lines = lyrics.split('\n')
    return sum(1 for line in lines if not line.strip())

def count_total_lines(lyrics):
    if pd.isna(lyrics):
        return 0
    return len(lyrics.split('\n'))

computed_sentences = df['lyrics'].apply(count_sentences_in_lyrics)
empty_lines = df['lyrics'].apply(count_empty_lines)
total_lines = df['lyrics'].apply(count_total_lines)

mismatches_sentences = df[df['n_sentences'] != computed_sentences]

print(f"matching rows: {len(df) - len(mismatches_sentences)}")
print(f"mismatching rows: {len(mismatches_sentences)}\n")

comparison_display = pd.DataFrame({
    'title': df['title'],
    'stored_n_sentences': df['n_sentences'],
    'computed_non_empty': computed_sentences,
    'empty_lines': empty_lines,
    'total_lines': total_lines,
    'diff': computed_sentences - df['n_sentences']
})
display(comparison_display.head(10))

matching rows: 4773
mismatching rows: 6393



,title,stored_n_sentences,computed_non_empty,empty_lines,total_lines,diff
0,​polka 2 :-/,102,98,5,103,-4
1,POLKA,56,56,5,61,0
2,​britney ;-),88,87,7,94,-1
3,CEO,37,37,4,41,0
4,LONDRA,48,48,6,54,0
5,BOHEME,62,62,8,70,0
6,LOBBY WAY,82,80,6,86,-2
7,SLATT,82,77,8,85,-5
8,MADE IN ITALY,82,82,8,90,0
9,ROSE & ROVI,50,49,3,52,-1


even if there are a lot of matching counting the lyric lines, and the mismatch looks close to the stored n_sentences value, we saw from the lyrics analysis that there are tracks stored in single lines, so it's not possible to count the number of sentences using the new lines. furthermore, there are many tracks containing more sentences for each line. let's try using a nlp library called spaCy:

In [56]:
def count_sentences_spacy(lyrics, lang, stored_n_sentences):
    if pd.isna(lyrics):
        return 0
    #only process italian and english (only two models used in the analysis)
    if lang not in ['it', 'en']:
        #return NA for missing language, otherwise keep stored value
        return 0 if pd.isna(lang) else stored_n_sentences
    
    nlp = models_tokenizers[lang]
    doc = nlp(lyrics)
    return len(list(doc.sents))

computed_sentences_spacy = df.apply(
    lambda row: count_sentences_spacy(row['lyrics'], row['language'], row['n_sentences']), 
    axis=1
)

#compare with stored values (only for rows where we computed sentences)
computed_mask = df['language'].isin(['it', 'en']) if 'language' in df.columns else pd.Series([True] * len(df))
mismatches_sentences_spacy = df[computed_mask & (df['n_sentences'] != computed_sentences_spacy)]

print(f"Rows computed (it/en): {computed_mask.sum()}")
print(f"Rows kept as stored (es/pt/fr/de/ru/bg): {(df['language'].notna() & ~computed_mask).sum()}")
print(f"Rows set to NA (missing language): {df['language'].isna().sum()}")
print(f"Matching rows: {computed_mask.sum() - len(mismatches_sentences_spacy)}")
print(f"Mismatching rows: {len(mismatches_sentences_spacy)}\n")

comparison_display = pd.DataFrame({
    'title': df['title'],
    'language': df['language'] if 'language' in df.columns else 'it',
    'stored_n_sentences': df['n_sentences'],
    'computed_spacy': computed_sentences_spacy,
    'computed_non_empty': computed_sentences,
    'diff_spacy': computed_sentences_spacy - df['n_sentences'],
    'diff_non_empty': computed_sentences - df['n_sentences']
})

# Sort by absolute difference in descending order and display top 10
display(comparison_display[computed_mask].head(10))

# cases where computed_non_empty is 1
print("\nCases where computed_non_empty = 1 (single line lyrics):")
single_line_mask = computed_sentences == 1
single_line_display = comparison_display[single_line_mask & computed_mask][
    ['title', 'language', 'stored_n_sentences', 'computed_spacy', 'computed_non_empty']
]
display(single_line_display.head(15))

Rows computed (it/en): 10983
Rows kept as stored (es/pt/fr/de/ru/bg): 55
Rows set to NA (missing language): 128
Matching rows: 96
Mismatching rows: 10887



,title,language,stored_n_sentences,computed_spacy,computed_non_empty,diff_spacy,diff_non_empty
0,​polka 2 :-/,it,102,6,98,-96,-4
1,POLKA,it,56,1,56,-55,0
2,​britney ;-),it,88,2,87,-86,-1
3,CEO,it,37,1,37,-36,0
4,LONDRA,it,48,1,48,-47,0
5,BOHEME,it,62,3,62,-59,0
6,LOBBY WAY,it,82,3,80,-79,-2
7,SLATT,it,82,4,77,-78,-5
8,MADE IN ITALY,it,82,1,82,-81,0
9,ROSE & ROVI,it,50,2,49,-48,-1



Cases where computed_non_empty = 1 (single line lyrics):


,title,language,stored_n_sentences,computed_spacy,computed_non_empty
497,Come Me,it,1,1,1
641,Escamotage,it,1,1,1
1777,Intro (Napolimanicomio),it,3,2,1
1929,Totòlude,it,8,8,1
1955,Dammi Ancora,it,11,11,1
1968,Cin Cin,it,1,1,1
1974,Tu Non Hai Mai,it,14,14,1
1975,Pronte A Tutto,it,3,2,1
1977,Non Dimentico Più,it,10,10,1
1978,Come Ti Senti?,it,6,6,1


since the puntuaction in some of this track is almost missing, spaCy can't obtain a real number of sentences. In order to fill n_sentences with a reasonable approssimation of the original column, lines will be count: if the lyric is in a single line (count == 1), then spaCy will be used

In [57]:
missing_n_sentences = ((df['n_sentences'].isna()) | (df['n_sentences'] == 0)) & df['lyrics'].notna()

print(f"rows with missing n_sentences but lyrics present: {missing_n_sentences.sum()}")

modified_rows = []

for idx in df[missing_n_sentences].index:
    lyrics = df.at[idx, 'lyrics']
    non_empty_count = count_sentences_in_lyrics(lyrics)
    
    #if only 1 non-empty line, use spaCy instead
    if non_empty_count == 1:
        lang = df.at[idx, 'language'] if 'language' in df.columns else 'it'
        if lang in ['it', 'en']:
            spacy_count = count_sentences_spacy(lyrics, lang, pd.NA)
            df.at[idx, 'n_sentences'] = spacy_count
            modified_rows.append({
                'index': idx,
                'title': df.at[idx, 'title'],
                'language': lang,
                'method': 'spacy',
                'n_sentences': spacy_count
            })
        else:
            df.at[idx, 'n_sentences'] = non_empty_count
            modified_rows.append({
                'index': idx,
                'title': df.at[idx, 'title'],
                'language': lang,
                'method': 'line_count',
                'n_sentences': non_empty_count
            })
    else:
        df.at[idx, 'n_sentences'] = non_empty_count
        modified_rows.append({
            'index': idx,
            'title': df.at[idx, 'title'],
            'language': df.at[idx, 'language'] if 'language' in df.columns else 'it',
            'method': 'line_count',
            'n_sentences': non_empty_count
        })

print(f"filled n_sentences for {missing_n_sentences.sum()} rows")
modified_df = pd.DataFrame(modified_rows)
display(modified_df.head(15))

rows with missing n_sentences but lyrics present: 70
filled n_sentences for 70 rows


,index,title,language,method,n_sentences
0,3667,Quelli che benpensano,it,line_count,79
1,3668,Autodafè,it,line_count,67
2,3669,Fight da faida,it,line_count,97
3,3670,Giù le mani da Caino,it,line_count,89
4,3671,Pedala,it,line_count,78
5,3672,Accendimi...,it,line_count,104
6,3673,Fili,it,line_count,69
7,3674,Rap Lamento,it,line_count,76
8,3675,Cali di tensione,it,line_count,94
9,3676,Cubetti tricolori,it,line_count,14


 ### n_tokens

In [58]:
invalid_elems = df[df['n_tokens'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['n_tokens'])

nan_indexes = df.index[df['n_tokens'].isna()].tolist()
print(f"number of missing values in n_tokens: {len(nan_indexes)}")

0        911.0
1        675.0
2        758.0
3        382.0
4        429.0
         ...  
11161    624.0
11162    602.0
11163    570.0
11164    304.0
11165    338.0
Name: n_tokens, Length: 11166, dtype: float64
number of missing values in n_tokens: 70


In [59]:
df['n_tokens'] = pd.to_numeric(df['n_tokens'], errors='coerce').astype('Int64')

def count_words(lyrics):
    if pd.isna(lyrics):
        return 0
    return len(lyrics.split()) #split by whitespace

computed_words = df['lyrics'].apply(count_words)

matches_words = (df['n_tokens'] == computed_words).sum()
matching_tracks = df[computed_words == df['n_tokens']].copy()
diff_words = computed_words - df['n_tokens']

print(f"matches: {matches_words}")
print(f"mean difference: {diff_words.mean():+.2f}")
print(f"Variance: {diff_words.var():+.2f}")
print(f"Standard deviation: {diff_words.std():+.2f}")

matching_tracks = matching_tracks[matching_tracks['lyrics'].notna()]

print(f"\nsome matching tracks:")
display_cols = ['title', 'n_tokens', 'lyrics']
display(matching_tracks[display_cols].head(20))

matches: 141
mean difference: -79.51
Variance: +2669.25
Standard deviation: +51.66

some matching tracks:


,title,n_tokens,lyrics
641,Escamotage,8,Biggie Paul sopra il beat sta in alto
1982,Intro,12,La Pina\nPina\nLa Pina\nPina\nLa Pina\nPina\nL...
2671,Nobiltà Di Strada (Intro),4,Inoki\nNobiltà di strada
3180,Star*,26,Senorita sai che sono un giovane re\n\nOcchi a...
3484,Cantico Errante Di Due D.J. Notturni,14,Questo è il nostro stile\n\nMica male\nDue\nDJ...
4012,Prologo,32,Prima che il gallo canti due volte\nTu avrai n...
4484,Unknown 8,57,Solo paranoie\nPrima o poi so che tornerai\nSo...
5149,LA CADUTA DELL'ANGELO - Intro,18,Mi ero svegliata in un posto che non conoscevo...
8162,Intro,27,La strada è la mia seconda madre\nRappresent C...
8854,Non esistono risposte intelligenti,3,A domande stupide


the number of tokens doesn't seem to count the tokens only by splitting by whitespace. From the mean difference, we can notice that there are missing tokens that are not getting counted. By analyzing some of the lyrics, all seem to have in common to be short (mostly errors) and with no punctuation, so i'll try to include it in the token count

In [60]:
def count_tokens_split_punctuation(lyrics):
    if pd.isna(lyrics):
        return 0
    import re
    #split on word boundaries - keeps alphanumeric sequences separate from punctuation
    # \w+ -> word characters (letters, numbers); [^\w\s] ->  punctuation (not words and not separators);
    tokens = re.findall(r'\w+|[^\w\s]', lyrics, flags=re.UNICODE)
    return len(tokens)

computed_split_punct = df['lyrics'].apply(count_tokens_split_punctuation)

matches_words = (df['n_tokens'] == computed_split_punct).sum()
matching_tracks = df[computed_split_punct == df['n_tokens']].copy()
diff_words = computed_split_punct - df['n_tokens']

print(f"matches: {matches_words}")
print(f"mean difference: {diff_words.mean():+.2f}")
print(f"Variance: {diff_words.var():+.2f}")
print(f"Standard deviation: {diff_words.std():+.2f}")

matching_tracks = matching_tracks[matching_tracks['lyrics'].notna()]

print(f"\nsome matching tracks:")
display_cols = ['title', 'n_tokens', 'lyrics']
display(matching_tracks[display_cols].head(20))

print("\nsome mismatching tracks:")
mismatch_tracks = df[computed_split_punct != df['n_tokens']].copy()
mismatch_tracks['computed_split_punct'] = computed_split_punct[mismatch_tracks.index]
mismatch_tracks['diff'] = mismatch_tracks['computed_split_punct'] - mismatch_tracks['n_tokens']
display_cols = ['title', 'n_tokens', 'computed_split_punct', 'diff', 'lyrics']
display(mismatch_tracks[display_cols].head(10))

matches: 592
mean difference: +3.49
Variance: +651.01
Standard deviation: +25.51

some matching tracks:


,title,n_tokens,lyrics
50,Blackout,241,"Chiusi dentro una stanza senza guardare l'ora,..."
62,Ritratti dell'odio,805,"Kranyo, ah, vai\n\nMi siedo e ripenso ""Mah, st..."
82,Chicas,333,Tutte matte le mie amiche\nTutte matte le mie ...
113,Mood Blu,396,"Tra tutte le scuse che mi racconto, mi circond..."
153,CI SARÒ,386,"Voglio guadagnarmi da vivere, ma non vivere pe..."
167,non so chi ha creato il mondo,45,Questa donna che riflette il cielo\nChe fa azz...
176,WNDLST Freestyle,219,"Eh, okay\nNon sono uno che parla tanto, ma sto..."
196,Omega (Outro),472,"Tranne una volta, lei ricorda bene\nSputo vele..."
235,BLESS SU BLESS,511,(Vedo bless) Da quando ero nella culla\n(Chied...
269,Cazzeggio Nella BV3 House,112,Probabilmente si tratta di un provino di BV3 s...



some mismatching tracks:


,title,n_tokens,computed_split_punct,diff,lyrics
0,​polka 2 :-/,911,945,34,"Oplà, ah\nBdope, chiama due b—\n\nMi candidere..."
1,POLKA,675,733,58,"Greg Willen, non dormire\n(Brr-poh)\n\nT-T-Tro..."
2,​britney ;-),758,792,34,"Mothz\nYeah, yeah, yeah-yeah\nBdope, chiama du..."
3,CEO,382,390,8,Designer sui vestiti penso di essere un outlet...
4,LONDRA,429,438,9,"Bdope (Yeah)\n\nVuole solo me, non fare la gel..."
5,BOHEME,387,456,69,Uh-uh\nAh-li-ah\nUh-uh\n\nSenti la pioggia che...
6,LOBBY WAY,680,671,-9,"Dback in your head, motherfuckers\nBdope, chia..."
7,SLATT,657,697,40,"Bdope, chiama due bitches\nPe-pe-pe-pe-pe-pew\..."
8,MADE IN ITALY,584,672,88,"Damn (Shlet)\nAhah, Rosa\nBdope, chiama due –\..."
9,ROSE & ROVI,589,627,38,"Cambiare, sì, ma non è facile (Pow)\nMi bacia ..."


this produces a result similar to the one in the dataset (with even long songs matching), but it's hard to say which are the chars that counted as tokens and which are not. let's see the results of a nlp library (spaCy):

In [61]:
def analyze_n_tokens_nlp(lyrics, lang, models_tokenizers):
    if pd.isna(lyrics):
        return 0
    nlp = models_tokenizers.get(lang, models_tokenizers['it']) #it as default
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", UserWarning)
        doc = nlp(lyrics)
    return len([t for t in doc if not t.is_space])

print("Starting spaCy tokenization")

df['real_n_tokens'] = df.apply(
    lambda row: analyze_n_tokens_nlp(row['lyrics'], row['language'], models_tokenizers),
    axis=1
)

n_tokens_stored = pd.to_numeric(df['n_tokens'], errors='coerce').astype('Int64')
real_n_tokens_nullable = pd.to_numeric(df['real_n_tokens'], errors='coerce').astype('Int64')

matches = (n_tokens_stored.fillna(-1) == real_n_tokens_nullable.fillna(-1)).sum()
total_rows = len(df)
mismatches = total_rows - matches

diff_tokens = real_n_tokens_nullable.astype('float') - n_tokens_stored.astype('float')
mean_diff = diff_tokens.mean()

print(f"Total rows processed: {total_rows}")
print(f"Matching rows (stored 'n_tokens' == 'real_n_tokens'): {matches}")
print(f"Mismatching rows: {mismatches}")
print(f"Mean difference (real_n_tokens - n_tokens): {mean_diff:+.2f}")
print(f"Variance: {diff_tokens.var():+.2f}")
print(f"Standard deviation: {diff_tokens.std():+.2f}")

mismatch_tracks = df[n_tokens_stored != real_n_tokens_nullable].copy()
mismatch_tracks['diff'] = real_n_tokens_nullable[mismatch_tracks.index] - n_tokens_stored[mismatch_tracks.index]
display_cols = ['title', 'language', 'n_tokens', 'real_n_tokens', 'diff', 'lyrics']
display(mismatch_tracks[display_cols].head(10))

Starting spaCy tokenization
Total rows processed: 11166
Matching rows (stored 'n_tokens' == 'real_n_tokens'): 340
Mismatching rows: 10826
Mean difference (real_n_tokens - n_tokens): -8.39
Variance: +508.00
Standard deviation: +22.54


,title,language,n_tokens,real_n_tokens,diff,lyrics
0,​polka 2 :-/,it,911,936,25,"Oplà, ah\nBdope, chiama due b—\n\nMi candidere..."
1,POLKA,it,675,674,-1,"Greg Willen, non dormire\n(Brr-poh)\n\nT-T-Tro..."
2,​britney ;-),it,758,784,26,"Mothz\nYeah, yeah, yeah-yeah\nBdope, chiama du..."
4,LONDRA,it,429,437,8,"Bdope (Yeah)\n\nVuole solo me, non fare la gel..."
5,BOHEME,it,387,454,67,Uh-uh\nAh-li-ah\nUh-uh\n\nSenti la pioggia che...
6,LOBBY WAY,it,680,670,-10,"Dback in your head, motherfuckers\nBdope, chia..."
7,SLATT,it,657,689,32,"Bdope, chiama due bitches\nPe-pe-pe-pe-pe-pew\..."
8,MADE IN ITALY,it,584,660,76,"Damn (Shlet)\nAhah, Rosa\nBdope, chiama due –\..."
9,ROSE & ROVI,it,589,621,32,"Cambiare, sì, ma non è facile (Pow)\nMi bacia ..."
10,POLKA 3,it,487,484,-3,"Bdope, (Mmh, yeah) chiama due —\n\nBaciami in ..."


a simple regex seems to be more accurate (more matches and closer mean difference) compared to the spaCy tokenization. It's also faster to use regex compared to a nlp library, which is why the regex will be used on the missing values.

In [62]:
missing_n_tokens = ((df['n_tokens'].isna()) | (df['n_tokens'] == 0)) & df['lyrics'].notna()

print(f"rows with missing n_tokens but lyrics present: {missing_n_tokens.sum()}")

modified_rows = []

for idx in df[missing_n_tokens].index:
    lyrics = df.at[idx, 'lyrics']
    regex_token_count = count_tokens_split_punctuation(lyrics)
    
    df.at[idx, 'n_tokens'] = regex_token_count
    
    modified_rows.append({
        'index': idx,
        'title': df.at[idx, 'title'],
        'n_tokens_filled': regex_token_count
    })

print(f"filled n_tokens for {len(modified_rows)} rows.")

modified_df = pd.DataFrame(modified_rows)
display(modified_df.head(15))

rows with missing n_tokens but lyrics present: 70
filled n_tokens for 70 rows.


,index,title,n_tokens_filled
0,3667,Quelli che benpensano,738
1,3668,Autodafè,510
2,3669,Fight da faida,851
3,3670,Giù le mani da Caino,748
4,3671,Pedala,565
5,3672,Accendimi...,932
6,3673,Fili,444
7,3674,Rap Lamento,635
8,3675,Cali di tensione,845
9,3676,Cubetti tricolori,134


### tokens_per_sent

In [63]:
invalid_elems = df[df['tokens_per_sent'].apply(check_type_validity, expected_type=int)]
print(invalid_elems['tokens_per_sent'])

nan_indexes = df.index[df['tokens_per_sent'].isna()].tolist()
print(f"number of missing values in tokens_per_sent: {len(nan_indexes)}")

0         8.931373
1        12.053571
2         8.613636
3        10.324324
4         8.937500
           ...    
11161     8.320000
11162     7.921053
11163     7.702703
11164     9.500000
11165     7.511111
Name: tokens_per_sent, Length: 11166, dtype: float64
number of missing values in tokens_per_sent: 70


In [64]:
expected_tokens_per_sent = df['n_tokens'] / df['n_sentences']

tolerance = 1e-6
matches = (abs(df['tokens_per_sent'] - expected_tokens_per_sent) <= tolerance) | (df['tokens_per_sent'].isna() & expected_tokens_per_sent.isna())

total_rows = len(df)
num_matches = matches.sum()
num_mismatches = total_rows - num_matches

print(f"Matching rows (correct n_tokens over n_sentences): {num_matches}")
print(f"Mismatching rows: {num_mismatches}")

Matching rows (correct n_tokens over n_sentences): 10968
Mismatching rows: 198


this column is correctly computed, the 198 missing values are given by the missing 76 values of n_tokens and n_sentences + 128 NaN values. let's fill the missing values:

In [65]:
mask_to_fill = (
    (df['tokens_per_sent'].isna()) | (df['tokens_per_sent'] == 0) &
    df['n_tokens'].notna() &
    df['n_sentences'].notna() &
    (df['n_sentences'] != 0)
)

indices_to_fill = df[mask_to_fill].index

n_tokens = df.loc[indices_to_fill, 'n_tokens']
n_sentences = df.loc[indices_to_fill, 'n_sentences']

new_values = n_tokens / n_sentences
        
df.loc[indices_to_fill, 'tokens_per_sent'] = new_values

print(f"successfully filled {len(indices_to_fill)} rows")

successfully filled 70 rows


### char_per_tok

In [66]:
invalid_elems = df[df['char_per_tok'].apply(check_type_validity, expected_type=float)]
print(invalid_elems['char_per_tok'])

nan_indexes = df.index[df['char_per_tok'].isna()].tolist()
print(f"number of missing values in char_per_tok: {len(nan_indexes)}")

Series([], Name: char_per_tok, dtype: float64)
number of missing values in char_per_tok: 70


In [67]:
def count_chars_no_whitespace(lyrics): #count characters excluding all whitespace
    if pd.isna(lyrics):
        return 0
    return len(re.sub(r'\s', '', lyrics))

total_chars_no_whitespace = df['lyrics'].apply(count_chars_no_whitespace)

n_tokens_safe = df['n_tokens'].replace(0, np.nan)

expected_cpt = total_chars_no_whitespace / n_tokens_safe

tolerance = 1e-6
matches = (abs(df['char_per_tok'] - expected_cpt) <= tolerance) | \
          (df['char_per_tok'].isna() & expected_cpt.isna())

num_matches = matches.sum()
num_mismatches = len(df) - num_matches

diff_series = (df['char_per_tok'] - expected_cpt).dropna()
mean_diff = diff_series.mean()
variance_diff = diff_series.var()
std_diff = diff_series.std()

print(f"matching rows: {num_matches}")
print(f"mismatching rows: {num_mismatches}")
print(f"Mean difference: {mean_diff:.6f}")
print(f"Variance: {variance_diff:.6f}")
print(f"Standard deviation: {std_diff:.6f}")
print(f"Mean absolute difference: {diff_series.abs().mean():.6f}\n")


if num_matches > 0:
    print("some matches:")

    has_lyrics = df['lyrics'].notna() & (df['lyrics'].str.strip() != "")
    matching_df = df[matches & has_lyrics].copy()
    matching_display = pd.DataFrame({
        'title': matching_df['title'],
        'computed_cpt': expected_cpt[matching_df.index],
        'n_tokens': matching_df['n_tokens'],
        'chars_no_ws': total_chars_no_whitespace[matching_df.index],
        'lyrics': matching_df['lyrics'],
    })
    display(matching_display.head(20))

if num_mismatches > 0:
    print("some mismatches:")
    
    mismatches_df = df[~matches].copy()
    mismatches_display = pd.DataFrame({
        'title': mismatches_df['title'],
        'stored_cpt': mismatches_df['char_per_tok'],
        'computed_cpt': expected_cpt[mismatches_df.index],
        'diff': (mismatches_df['char_per_tok'] - expected_cpt[mismatches_df.index]).abs()
    })
    
    display(mismatches_display.head(10))

matching rows: 45
mismatching rows: 11121
Mean difference: 0.320687
Variance: 0.039178
Standard deviation: 0.197934
Mean absolute difference: 0.321348

some matches:


,title,computed_cpt,n_tokens,chars_no_ws,lyrics
641,Escamotage,3.75,8,30,Biggie Paul sopra il beat sta in alto
758,Intro (Dinamite),4.4,5,22,violence\n rastaman\nI don't
964,Cose Non Sane (Interlude),3.962963,27,107,Facevamo l'amore per non litigare\nPer avere f...
1155,Balb & Tando,4.52459,122,552,Per le ragazze sono negato incapace impacciato...
1432,Ma vaffanculo,4.473684,19,85,Lasciami andare a dormire\nCome se avessi la s...
1927,Interlude 1 (Mi Fist),4.0,30,120,I just can't understand\nI just can't understa...
1982,Intro,3.333333,12,40,La Pina\nPina\nLa Pina\nPina\nLa Pina\nPina\nL...
2671,Nobiltà Di Strada (Intro),5.0,4,20,Inoki\nNobiltà di strada
3180,Star*,4.423077,26,115,Senorita sai che sono un giovane re\n\nOcchi a...
3202,Non Voglio Morire Qui,4.014634,205,823,Linee scritte sulla pelle\nCorri fino a che po...


some mismatches:


,title,stored_cpt,computed_cpt,diff
0,​polka 2 :-/,4.170455,3.45225,0.718204
1,POLKA,4.280851,3.285926,0.994925
2,​britney ;-),4.075251,3.411609,0.663641
3,CEO,4.023881,3.623037,0.400844
4,LONDRA,3.922857,3.379953,0.542904
5,BOHEME,4.160606,3.697674,0.462932
6,LOBBY WAY,3.817680,3.251471,0.566209
7,SLATT,4.031373,3.343988,0.687385
8,MADE IN ITALY,3.935841,3.263699,0.672142
9,ROSE & ROVI,3.851508,3.086587,0.764921


In [68]:
mask_to_fill = (
    (df['char_per_tok'].isna()) | (df['char_per_tok'] == 0) &
    df['n_tokens'].notna() &
    df['n_sentences'].notna() &
    (df['n_tokens'] != 0)
)

indices_to_fill = df[mask_to_fill].index

chars_no_ws = df.loc[indices_to_fill, 'lyrics'].apply(count_chars_no_whitespace)
n_tokens = df.loc[indices_to_fill, 'n_tokens']

new_values = chars_no_ws / n_tokens

#fill in the missing values
df.loc[indices_to_fill, 'char_per_tok'] = new_values

print(f"successfully filled {len(indices_to_fill)} rows")

successfully filled 70 rows


### lexical_density

In [69]:
invalid_elems = df[df['lexical_density'].apply(check_type_validity, expected_type=float)]
print(invalid_elems['lexical_density'])

nan_indexes = df.index[df['lexical_density'].isna()].tolist()
print(f"number of missing values in lexical_density: {len(nan_indexes)}")

Series([], Name: lexical_density, dtype: float64)
number of missing values in lexical_density: 70


In [70]:
invalid_density = df[(df['lexical_density'] < 0) | (df['lexical_density'] > 1)]
print(f"rows with invalid density (not in [0, 1] range): {len(invalid_density)}")

def compute_lexical_density_regex(lyrics):
    if pd.isna(lyrics):
        return 0
    
    all_tokens = re.findall(r"[^\W\d_]+(?:['’-][^\W\d_]+)*", lyrics.lower(), flags=re.UNICODE) #finds both words and numbers
    
    all_words = all_tokens #filter out any tokens that are just numbers
    
    total_words = len(all_words)
    if total_words == 0:
        return 0
    
    unique_words = len(set(all_words)) #count of unique words
    
    return unique_words / total_words

computed_ld_regex = df['lyrics'].apply(compute_lexical_density_regex)

tolerance = 1e-3
matches_regex = (abs(df['lexical_density'] - computed_ld_regex) <= tolerance) | \
               (df['lexical_density'].isna() & computed_ld_regex.isna())

num_matches_regex = matches_regex.sum()
total_rows = len(df)

diff_ld = computed_ld_regex - df['lexical_density']
mean_diff_ld = diff_ld.mean()

print("\nVerifying calculation (Regex: unique_alpha_words / total_alpha_words):")
print(f"Total rows: {total_rows}")
print(f"Matching rows: {num_matches_regex}")
print(f"Mismatching rows: {total_rows - num_matches_regex}")
print(f"mean difference: {mean_diff_ld:+.6f}")
print(f"Variance: {diff_ld.var():.6f}")
print(f"Standard deviation: {diff_ld.std():.6f}")

# 5. Display Mismatches
if (total_rows - num_matches_regex) > 0:
    mismatches = df[~matches_regex].copy()
    mismatches_display = pd.DataFrame({
        'title': mismatches['title'],
        'language': mismatches['language'],
        'stored_lexical_density': mismatches['lexical_density'],
        'computed_ld_regex': computed_ld_regex[mismatches.index]
    })
    print("\nSample mismatches (Regex):")
    display(mismatches_display.head(10))

rows with invalid density (not in [0, 1] range): 0

Verifying calculation (Regex: unique_alpha_words / total_alpha_words):
Total rows: 11166
Matching rows: 213
Mismatching rows: 10953
mean difference: -0.023536
Variance: 0.013457
Standard deviation: 0.116005

Sample mismatches (Regex):


,title,language,stored_lexical_density,computed_ld_regex
0,​polka 2 :-/,it,0.575284,0.536296
1,POLKA,it,0.648936,0.476852
2,​britney ;-),it,0.556856,0.473684
3,CEO,it,0.534328,0.595541
4,LONDRA,it,0.491429,0.494083
5,BOHEME,it,0.506061,0.509434
6,LOBBY WAY,it,0.438306,0.382129
7,SLATT,it,0.535294,0.469636
8,MADE IN ITALY,it,0.537611,0.370098
9,ROSE & ROVI,it,0.554524,0.265060


even if the mean result is close, the variance seems high. let's see the results of a spaCy:

In [71]:
def analyze_lexical_density_nlp(lyrics, lang, models_tokenizers, stored_lexical_density):
    if pd.isna(lyrics):
        return 0 #return 0 for missing lyrics

    #return the stored value for non supported languages
    if lang not in ['it', 'en']:
        return stored_lexical_density

    nlp = models_tokenizers[lang]

    with warnings.catch_warnings():
        warnings.simplefilter("ignore", UserWarning)
        doc = nlp(lyrics)
    
    all_alpha_words = [t.lower_ for t in doc if t.is_alpha]
    total_alpha = len(all_alpha_words)
    
    if total_alpha == 0:
        return 0
    else:
        unique_alpha = len(set(all_alpha_words))
        return unique_alpha / total_alpha

df['nlp_lexical_density'] = df.apply(
    lambda row: analyze_lexical_density_nlp(
        row['lyrics'], 
        row['language'], 
        models_tokenizers,
        row['lexical_density']
    ),
    axis=1
)

computed_mask = df['language'].isin(['it', 'en'])

computed_stored_ld = df.loc[computed_mask, 'lexical_density']
computed_nlp_ld = df.loc[computed_mask, 'nlp_lexical_density']

tolerance = 1e-3
matches_ld = (abs(computed_stored_ld - computed_nlp_ld) <= tolerance) | \
             (computed_stored_ld.isna() & computed_nlp_ld.isna())

num_computed_rows = computed_mask.sum()
num_matches_ld = matches_ld.sum()
mismatches = num_computed_rows - num_matches_ld

diff_ld_computed = computed_nlp_ld - computed_stored_ld

valid_diffs = diff_ld_computed.dropna() 

mean_diff_ld = valid_diffs.mean()
variance_diff = valid_diffs.var()
std_diff = valid_diffs.std()

total_rows = len(df)

print(f"rows in DataFrame: {total_rows}")
print(f"rows computed (it/en): {num_computed_rows}")
print(f"rows kept as stored (other langs + 0s): {total_rows - num_computed_rows}")

print(f"matching rows: {num_matches_ld}")
print(f"mismatching rows: {mismatches}")

print(f"Mean difference (over {len(valid_diffs)} non-NA rows): {mean_diff_ld:+.6f}")
print(f"Variance: {variance_diff:.6f}")
print(f"Standard deviation: {std_diff:.6f}\n")

stored_density = df['lexical_density']
computed_density_spacy = df['nlp_lexical_density']

diff_spacy = computed_density_spacy - stored_density
diff_regex = computed_ld_regex - stored_density

comparison_display = pd.DataFrame({
    'title': df['title'],
    'language': df['language'] if 'language' in df.columns else 'it',
    'stored_lexical_density': stored_density,
    'computed_spacy': computed_density_spacy,
    'computed_regex': computed_ld_regex,
    'diff_spacy': diff_spacy,
    'diff_regex': diff_regex
})

display(comparison_display[computed_mask].head(10))

valid_comparison_mask = (
    computed_mask & 
    stored_density.notna() & 
    computed_density_spacy.notna() & 
    computed_ld_regex.notna()
)

abs_diff_spacy = abs(diff_spacy[valid_comparison_mask])
abs_diff_regex = abs(diff_regex[valid_comparison_mask])

closer_to_spacy = (abs_diff_spacy < abs_diff_regex).sum()
closer_to_regex = (abs_diff_regex < abs_diff_spacy).sum()
ties = (abs_diff_spacy == abs_diff_regex).sum()

print(f"rows where spaCy density is closer:   {closer_to_spacy}")
print(f"rows where regex density is closer:  {closer_to_regex}")
print(f"rows with equal difference (ties): {ties}")

rows in DataFrame: 11166
rows computed (it/en): 10983
rows kept as stored (other langs + 0s): 183
matching rows: 76
mismatching rows: 10907
Mean difference (over 10913 non-NA rows): -0.031116
Variance: 0.013596
Standard deviation: 0.116603



,title,language,stored_lexical_density,computed_spacy,computed_regex,diff_spacy,diff_regex
0,​polka 2 :-/,it,0.575284,0.509380,0.536296,-0.065905,-0.038988
1,POLKA,it,0.648936,0.444954,0.476852,-0.203982,-0.172084
2,​britney ;-),it,0.556856,0.447635,0.473684,-0.109221,-0.083172
3,CEO,it,0.534328,0.572327,0.595541,0.037999,0.061213
4,LONDRA,it,0.491429,0.474138,0.494083,-0.017291,0.002654
5,BOHEME,it,0.506061,0.444134,0.509434,-0.061927,0.003373
6,LOBBY WAY,it,0.438306,0.376894,0.382129,-0.061412,-0.056176
7,SLATT,it,0.535294,0.444444,0.469636,-0.090850,-0.065658
8,MADE IN ITALY,it,0.537611,0.309979,0.370098,-0.227632,-0.167513
9,ROSE & ROVI,it,0.554524,0.240826,0.265060,-0.313699,-0.289464


rows where spaCy density is closer:   3641
rows where regex density is closer:  6157
rows with equal difference (ties): 1115


they both have similar results, so for performances reasons the regex version will be instead to fill the missing rows.

In [72]:
mask_to_fill = (
    (df['lexical_density'].isna()) | (df['lexical_density'] == 0) &
    computed_ld_regex.notna()
)

num_filled = mask_to_fill.sum()

new_values = computed_ld_regex[mask_to_fill]

df.loc[mask_to_fill, 'lexical_density'] = new_values

print(f"successfully filled {num_filled} 'lexical_density' rows")

successfully filled 198 'lexical_density' rows


### avg_token_per_clause

In [73]:
invalid_elems = df[df['avg_token_per_clause'].apply(check_type_validity, expected_type=float)]
print(invalid_elems['avg_token_per_clause'])

nan_indexes = df.index[df['avg_token_per_clause'].isna()].tolist()
print(f"number of missing values in avg_token_per_clause: {len(nan_indexes)}")

Series([], Name: avg_token_per_clause, dtype: float64)
number of missing values in avg_token_per_clause: 70


In [74]:
def analyze_avg_token_per_clause(lyrics, lang, models_pos_taggers, stored_avg_token_per_clause):
    if pd.isna(lyrics):
        return 0

    if lang not in ['it', 'en']:
        return stored_avg_token_per_clause
    
    nlp = models_pos_taggers[lang]
    
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", UserWarning)
        doc = nlp(lyrics)
        
    total_non_punct_tokens = len([t for t in doc if not t.is_punct and not t.is_space])
    num_predicates = len([t for t in doc if t.pos_ in ('VERB', 'AUX')])
    
    if num_predicates == 0:
        return 0
    else:
        return total_non_punct_tokens / num_predicates


df['cmp_avg_token_per_clause'] = df.apply(
    lambda row: analyze_avg_token_per_clause(row['lyrics'], row['language'], models_pos_taggers, row['avg_token_per_clause']),
    axis=1
)

computed_mask = df['language'].isin(['it', 'en'])

computed_stored = df.loc[computed_mask, 'avg_token_per_clause']
computed_real = df.loc[computed_mask, 'cmp_avg_token_per_clause']

tolerance = 1e-3
matches_clause = (abs(computed_stored - computed_real) <= tolerance) | \
                 (computed_stored.isna() & computed_real.isna())

num_computed_rows = computed_mask.sum()
num_matches_clause = matches_clause.sum()
mismatches = num_computed_rows - num_matches_clause

diff_clause_computed = computed_real - computed_stored
    
valid_diffs = diff_clause_computed.dropna() 

mean_diff_clause = valid_diffs.mean()
variance_diff = valid_diffs.var()
std_diff = valid_diffs.std()

print(f"rows kept as stored (other langs + 0s): {total_rows - num_computed_rows}")
print(f"matching rows: {num_matches_clause}")
print(f"mismatching rows: {mismatches}")
print(f"mean difference (real - stored): {mean_diff_clause:+.6f}\n")
print(f"variance (real - stored): {variance_diff:.6f}")
print(f"std. Deviation (real - stored): {std_diff:.6f}\n")

if mismatches > 0:
        mismatch_indices = matches_clause[~matches_clause].index
        mismatches_df = df.loc[mismatch_indices].copy()
        mismatches_df['diff'] = diff_clause_computed[mismatch_indices]
        
        print("sample mismatches (from 'it'/'en' rows)")
        display(mismatches_df[['title', 'language', 'avg_token_per_clause', 'cmp_avg_token_per_clause', 'diff']].head(10))

rows kept as stored (other langs + 0s): 183
matching rows: 19
mismatching rows: 10964
mean difference (real - stored): -2.272159

variance (real - stored): 136.223407
std. Deviation (real - stored): 11.671478

sample mismatches (from 'it'/'en' rows)


,title,language,avg_token_per_clause,cmp_avg_token_per_clause,diff
0,​polka 2 :-/,it,8.133929,6.805825,-1.328103
1,POLKA,it,12.500000,7.387097,-5.112903
2,​britney ;-),it,8.422222,6.270833,-2.151389
3,CEO,it,6.701754,4.880597,-1.821157
4,LONDRA,it,8.411765,5.932203,-2.479561
5,BOHEME,it,6.450000,5.901639,-0.548361
6,LOBBY WAY,it,12.142857,8.265625,-3.877232
7,SLATT,it,10.265625,7.123288,-3.142337
8,MADE IN ITALY,it,11.018868,7.208955,-3.809913
9,ROSE & ROVI,it,9.349206,6.138889,-3.210317


here we don't have much choise other than relying on spaCy to identify the clauses.

In [75]:
mask_to_fill = (
    (df['avg_token_per_clause'].isna()) | (df['avg_token_per_clause'] == 0) &
    df['cmp_avg_token_per_clause'].notna() &
    df['cmp_avg_token_per_clause'] != 0
)

num_filled = mask_to_fill.sum()

new_values = df.loc[mask_to_fill, 'cmp_avg_token_per_clause']

df.loc[mask_to_fill, 'avg_token_per_clause'] = new_values

print(f"Successfully filled {num_filled} 'avg_token_per_clause' rows.")

Successfully filled 81 'avg_token_per_clause' rows.
